### PR TITLE
Fs backtrace tweak

### PR DESCRIPTION
**Pull request by XuNeo**
## Summary

~Add option to dump fdinfo when assert happens, as required in~

`file_ioctl` is required to get file path, thus not available in assert. Currently file list is only dumped on task exit.

https://github.com/apache/nuttx/pull/12532#issuecomment-2179500766

nxsched_dumponexit has been moved forward to task_exithook, in order to access the `group` before it's destroyed.

Example of crash log:
```
nsh> mw -1
_assert: Current Version: NuttX  12.5.1 a8d7f155b8 Jun 28 2024 23:19:10 arm64
_assert: Assertion failed panic: at file: /home/neo/projects/nuttx/nuttx/arch/arm64/src/common/arm64_fatal.c:375 task: nsh_main process: nsh_main 0x402a10a8
up_dump_register: stack = 0x4040a230
up_dump_register: x0:   0xffffffffffffffff  x1:   0x0
up_dump_register: x2:   0x4040a300          x3:   0x0
up_dump_register: x4:   0x402a7934          x5:   0xa820a8a8aaa8aaaa
up_dump_register: x6:   0x80adb1808080      x7:   0xfefefefefeff766c
up_dump_register: x8:   0x7f7f7f7f7f7f7f7f  x9:   0x7f7f7f7f7f7f7f7f
up_dump_register: x10:  0x101010101010101   x11:  0x10
up_dump_register: x12:  0x1                 x13:  0xffffffffffffffff
up_dump_register: x14:  0xffff000000000000  x15:  0xffffffffffffffff
up_dump_register: x16:  0x0                 x17:  0x0
up_dump_register: x18:  0x0                 x19:  0x1
up_dump_register: x20:  0x403f1038          x21:  0x0
up_dump_register: x22:  0x0                 x23:  0x0
up_dump_register: x24:  0x0                 x25:  0x0
up_dump_register: x26:  0x0                 x27:  0x0
up_dump_register: x28:  0x0                 x29:  0x4040a360
up_dump_register: x30:  0x402b2298
up_dump_register:
up_dump_register: STATUS Registers:
up_dump_register: SPSR:      0x80000005
up_dump_register: ELR:       0x402b22c4
up_dump_register: SP_EL0:    0x4040a640
up_dump_register: SP_ELX:    0x4040a360
up_dump_register: TPIDR_EL0: 0x40406220
up_dump_register: TPIDR_EL1: 0x40406220
up_dump_register: EXE_DEPTH: 0x0
backtrace:
[ 2] [<0x402b22c4>] cmd_mw+0x58/0x100
[ 2] [<0x402ac514>] nsh_command+0x10c/0x11c
[ 2] [<0x402a91a0>] nsh_execute+0x390/0x40c
[ 2] [<0x402ab2a0>] nsh_parse_command+0x4d4/0x524
[ 2] [<0x402ab3a0>] nsh_parse+0xb0/0x1b4
[ 2] [<0x402a7f70>] nsh_session+0x284/0x298
[ 2] [<0x402a77b8>] nsh_consolemain+0x74/0x94
[ 2] [<0x402a1100>] nsh_main+0x58/0x84
[ 2] [<0x402a4208>] nxtask_startup+0x58/0x5c
[ 2] [<0x4029e59c>] nxtask_start+0xe8/0xf0
dump_tasks:    PID GROUP PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
dump_tasks:   ----   --- --- -------- ------- --- ------- ---------- ---------------- 0x40402000      4096       544    13.2%    irq
dump_task:       0     0   0 FIFO     Kthread - Ready              0000000000000000 0x40400010      8176      2080    25.4%    Idle_Task
dump_task:       1     0 192 RR       Kthread - Waiting Semaphore  0000000000000000 0x40404210      8112      1256    15.4%    hpwork 0x403fb2d0 0x403fb2f8
dump_task:       2     2 100 RR       Task    - Running            0000000000000000 0x404089b0      8144      4076    50.0%    nsh_main
backtrace:
[ 0] [<0x402a2350>] arch_cpu_idle+0x8/0xc
[ 0] [<0x402975d8>] nx_start+0x244/0x248
[ 0] [<0x402810e8>] arm64_boot_primary_c_routine+0x18/0x24
backtrace:
[ 1] [<0x402a1694>] sys_call2+0x24/0x2c
[ 1] [<0x402a170c>] up_switch_context+0x70/0x7c
[ 1] [<0x402991f8>] nxsem_wait+0x1f0/0x22c
[ 1] [<0x40299248>] nxsem_wait_uninterruptible+0x14/0x30
[ 1] [<0x4029b1a4>] work_thread+0x114/0x118
[ 1] [<0x4029e580>] nxtask_start+0xcc/0xf0
backtrace:
[ 2] [<0x402b22c4>] cmd_mw+0x58/0x100
[ 2] [<0x402ac514>] nsh_command+0x10c/0x11c
[ 2] [<0x402a91a0>] nsh_execute+0x390/0x40c
[ 2] [<0x402ab2a0>] nsh_parse_command+0x4d4/0x524
[ 2] [<0x402ab3a0>] nsh_parse+0xb0/0x1b4
[ 2] [<0x402a7f70>] nsh_session+0x284/0x298
[ 2] [<0x402a77b8>] nsh_consolemain+0x74/0x94
[ 2] [<0x402a1100>] nsh_main+0x58/0x84
[ 2] [<0x402a4208>] nxtask_startup+0x58/0x5c
[ 2] [<0x4029e59c>] nxtask_start+0xe8/0xf0
fd: 0  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x00000000402a2ee4 0x00000000402a3058 0x000000004029bb8c 0x0000000040297544 0x00000000402810e8
fd: 1  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028d84c 0x000000004028e18c 0x000000004028e1b4 0x000000004029bba8 0x0000000040297544 0x00000000402810e8
fd: 2  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028d84c 0x000000004028e18c 0x000000004028e1b4 0x000000004029bbb4 0x0000000040297544 0x00000000402810e8
fd: 0  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x00000000402a2ee4 0x00000000402a3058 0x000000004029bb8c 0x0000000040297544 0x00000000402810e8
fd: 1  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028d84c 0x000000004028e18c 0x000000004028e1b4 0x000000004029bba8 0x0000000040297544 0x00000000402810e8
fd: 2  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028d84c 0x000000004028e18c 0x000000004028e1b4 0x000000004029bbb4 0x0000000040297544 0x00000000402810e8
fd: 0  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028f1f8 0x000000004028e050 0x000000004029bd08 0x000000004029da10 0x000000004029a36c 0x000000004029a4d8 0x000000004029a5a8 0x0000000040297644 0x0000000040297698 0x00000000402976cc 0x0000000040297598 0x00000000402810e8
fd: 1  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028f1f8 0x000000004028e050 0x000000004029bd08 0x000000004029da10 0x000000004029a36c 0x000000004029a4d8 0x000000004029a5a8 0x0000000040297644 0x0000000040297698 0x00000000402976cc 0x0000000040297598 0x00000000402810e8
fd: 2  , oflags: 3       type: 1    pos: 0        path: /dev/console BACKTRACE: 0x000000004028f1f8 0x000000004028e050 0x000000004029bd08 0x000000004029da10 0x000000004029a36c 0x000000004029a4d8 0x000000004029a5a8 0x0000000040297644 0x0000000040297698 0x00000000402976cc 0x0000000040297598 0x00000000402810e8

```
## Impact

No

## Testing

qemu-armv8a:nsh with related options enabled.

### CODE REVIEW

File: fs/inode/fs_files.c [hunk 1]
```diff
@@ -308,6 +311,59 @@ void files_initlist(FAR struct filelist *list)
...
+        {
+          path[0] = '\0';
+        }
+
+#if CONFIG_FS_BACKTRACE > 0
+      backtrace_format(buf, sizeof(buf), filep->f_backtrace,
+                       CONFIG_FS_BACKTRACE);
+#endif
+
+      syslog(LOG_INFO, "fd: %-3d, oflags: %-7d type: %-4x pos: %-9ld"
```
Comment (anchao) [inline fs/inode/fs_files.c:353]: refine the format pattern:
```
pid fd oflags type pos path backtrace
** ** ** ** **
```

**Review by anchao — CHANGES_REQUESTED** [review]
Please do not merge this PR before complete the review

File: fs/inode/fs_files.c [hunk 2]
```diff
@@ -308,6 +311,59 @@ void files_initlist(FAR struct filelist *list)
...
+#endif
+
+      /* Is there an inode associated with the file descriptor? */
+
+      if (!filep || !filep->f_inode)
+        {
+          continue;
+        }
+
+      if (file_ioctl(filep, FIOC_FILEPATH, path) < 0)
```
[the remark this answers is not in the document]
Comment (anchao) [reply fs/inode/fs_files.c:343]: eg. 
The socket file object will hold the net_lock() in the file_ioctl(). Please verify any changes on a fully functional product.
https://github.com/apache/nuttx/blob/master/net/tcp/tcp_ioctl.c#L122

File: fs/inode/fs_files.c (hunk 1 above)
Comment (XuNeo) [reply fs/inode/fs_files.c]: Done. Example of outputs:
```
nsh> hello
Hello, World!!
tcb=0x403f93a0 name=Idle_Task, pid:0, priority=0 state=2 stack_alloc_ptr: 0x40400000, adj_stack_size: 8176
PID FD  FLAGS   TYPE POS       PATH           BACKTRACE
3   0   3       1    0         /dev/console   0x00000000402a32ac 0x00000000402a3420 0x000000004029bf54 0x0000000040297900 0x00000000402810e8
3   1   3       1    0         /dev/console   0x000000004028d870 0x000000004028e064 0x000000004028e08c 0x000000004029bf70 0x0000000040297900 0x00000000402810e8
3   2   3       1    0         /dev/console   0x000000004028d870 0x000000004028e064 0x000000004028e08c 0x000000004029bf7c 0x0000000040297900 0x00000000402810e8
```

File: include/execinfo.h [hunk 3]
```diff
@@ -27,15 +27,43 @@
...
+
+/* Format a backtrace into a buffer for dumping. */
+
+#define backtrace_format(buf, buflen, backtrace, depth)              \
+do                                                                   \
+  {                                                                  \
+    FAR const char *format = "%0*p ";                                \
+    int i;                                                           \
+                                                                     \
+    DEBUGASSERT(buflen >= BACKTRACE_BUFFER_SIZE(depth));             \
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply include/execinfo.h]: Done

File: include/execinfo.h [hunk 4]
```diff
@@ -27,15 +27,43 @@
...
+/* Buffer size needed to hold formatted `depth` backtraces */
+
+#define BACKTRACE_BUFFER_SIZE(d) (BACKTRACE_PTR_FMT_WIDTH * (d) + 1)
+
+#define backtrace(b, s) sched_backtrace(_SCHED_GETTID(), b, s, 0)
+#define dump_stack()    sched_dumpstack(_SCHED_GETTID())
+
+/* Format a backtrace into a buffer for dumping. */
+
+#define backtrace_format(buf, buflen, backtrace, depth)              \
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply include/execinfo.h]: Done

File: fs/vfs/fs_open.c [hunk 5]
```diff
@@ -350,6 +350,11 @@ int file_open(FAR struct file *filep, FAR const char *path, int oflags, ...)
   ret = file_vopen(filep, path, oflags, 0, ap);
   va_end(ap);
 
+  if (ret >= 0)
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply fs/vfs/fs_open.c]: Done

File: fs/inode/fs_files.c [hunk 6]
```diff
@@ -308,6 +311,59 @@ void files_initlist(FAR struct filelist *list)
...
+      FAR struct file *filep = files_fget(list, i);
+      char path[PATH_MAX];
+
+#if CONFIG_FS_BACKTRACE > 0
+      char buf[BACKTRACE_BUFFER_SIZE(CONFIG_FS_BACKTRACE)];
+#endif
+
+      /* Is there an inode associated with the file descriptor? */
+
+      if (!filep || !filep->f_inode)
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply fs/inode/fs_files.c]: Done

File: fs/procfs/fs_procfsproc.c [hunk 7]
```diff
@@ -1244,17 +1245,16 @@ static ssize_t proc_groupfd(FAR struct proc_file_s *procfile,
                             size_t buflen, off_t offset)
 {
   FAR struct task_group_s *group = tcb->group;
+  FAR struct file *filep;
+  char backtrace[BACKTRACE_BUFFER_SIZE(CONFIG_FS_BACKTRACE)];
```
[the remark this answers is not in the document]
Comment (XuNeo) [reply fs/procfs/fs_procfsproc.c]: I have picked your change and updated to the common backtrace_format function. Please help to review.

File: fs/inode/fs_files.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (XuNeo) [reply fs/inode/fs_files.c:351]: Let's remove the function to dump file list in assert until we find a better way.

**Review by anchao — COMMENTED** [review]
LGTM

**Review by anchao — APPROVED** [review]

File: include/nuttx/fs/fs.h [hunk 8]
```diff
@@ -865,15 +865,25 @@ int nx_umount2(FAR const char *target, unsigned int flags);
 void files_initlist(FAR struct filelist *list);
 
 /****************************************************************************
- * Name: files_releaselist
+ * Name: files_dumplist
```
Comment (xiaoxiang781216) [inline include/nuttx/fs/fs.h:865]: why change

File: include/nuttx/fs/fs.h [hunk 9]
```diff
@@ -865,15 +865,25 @@ int nx_umount2(FAR const char *target, unsigned int flags);
...
  *
  * Description:
- *   Release a reference to the file list
```
Comment (xiaoxiang781216) [inline include/nuttx/fs/fs.h:871]: revert

File: include/nuttx/fs/fs.h [hunk 10]
```diff
@@ -865,15 +865,25 @@ int nx_umount2(FAR const char *target, unsigned int flags);
...
  *
  ****************************************************************************/
 
 void files_releaselist(FAR struct filelist *list);
 
+/****************************************************************************
+ * Name: files_dumplist
+ *
+ * Description:
+ *   Release a reference to the file list
```
Comment (xiaoxiang781216) [inline include/nuttx/fs/fs.h:878]: wrong comment

File: include/nuttx/fs/fs.h (hunk 9 above)
Comment (XuNeo) [reply include/nuttx/fs/fs.h]: Done

File: include/nuttx/fs/fs.h (hunk 10 above)
Comment (XuNeo) [reply include/nuttx/fs/fs.h]: done

File: include/nuttx/fs/fs.h (hunk 8 above)
Comment (XuNeo) [reply include/nuttx/fs/fs.h:865]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**